File: generators/pixelate.cpp

// Part of the ProceduralTextureMaker project.
// http://github.com/johanokl/ProceduralTextureMaker
// Released under GPLv3.
// Johan Lindqvist (johan.lindqvist@gmail.com)

#include "pixelate.h"

PixelateTextureGenerator::PixelateTextureGenerator() {
   TextureGeneratorSetting width;
   width.defaultvalue = QVariant((double)5);
   width.name = "Block width (%)";
   width.description = "Width of each pixel block as a percentage of the texture width.";
   width.min = QVariant(0);
   width.max = QVariant(50);
   width.group = "size";
   width.id = "width";
   configurables.append(width);

   TextureGeneratorSetting height;
   height.defaultvalue = QVariant((double)5);
   height.name = "Block height (%)";
   height.description = "Height of each pixel block as a percentage of the texture height.";
   height.min = QVariant(0);
   height.max = QVariant(50);
   height.group = "size";
   height.id = "height";
   configurables.append(height);

   TextureGeneratorSetting offsetx;
   offsetx.defaultvalue = QVariant((double)0);
   offsetx.name = "Horizontal offset (%)";
   offsetx.description = "Moves the block grid horizontally across the texture.";
   offsetx.min = QVariant(-50);
   offsetx.max = QVariant(50);
   offsetx.id = "offsetx";
   configurables.append(offsetx);

   TextureGeneratorSetting offsety;
   offsety.defaultvalue = QVariant((double)0);
   offsety.name = "Vertical offset (%)";
   offsety.description = "Moves the block grid vertically across the texture.";
   offsety.min = QVariant(-50);
   offsety.max = QVariant(50);
   offsety.id = "offsety";
   configurables.append(offsety);
}
void PixelateTextureGenerator::generate(QSize size, TexturePixel* destimage,
                                        const QMap<QString, TextureImagePtr>& sourceimages,
                                        const TextureNodeSettings& settings) const {
   if (!destimage || !size.isValid()) {
      return;
   }
   int width = qMax(settings.value("width").toDouble() / 100 * size.width(), (double)1);
   int height = qMax(settings.value("height").toDouble() / 100 * size.height(), (double)1);
   int offsetx = settings.value("offsetx").toDouble() / 100 * size.width();
   int offsety = settings.value("offsety").toDouble() / 100 * size.height();
   if (!sourceimages.contains(QStringLiteral("Input"))) {
      memset(destimage, 0, size.width() * size.height() * sizeof(TexturePixel));
      return;
   }
   TexturePixel* sourceImage = sourceimages.value(QStringLiteral("Input")).data()->getData();
   int y = offsety - size.height();
   while (y < size.height()) {
      if (y + height > 0) {
         int x = offsetx - size.width();
         while (x < size.width()) {
            if (x + width > 0) {
               int totalPixels = 0;
               quint64 red = 0;
               quint64 green = 0;
               quint64 blue = 0;
               quint64 alpha = 0;
               for (int ypos = y; ypos < y + height; ypos++) {
                  int currY = ypos;
                  if (currY < 0) {
                     currY += size.height() * 5;
                  }
                  currY %= size.height();
                  currY *= size.width();
                  for (int xpos = x; xpos < x + width; xpos++) {
                     int currX = xpos;
                     if (currX < 0) {
                        currX += size.width() * 5;
                     }
                     currX %= size.width();
                     TexturePixel sourcePixel = sourceImage[currY + currX];
                     totalPixels++;
                     red += sourcePixel.r;
                     green += sourcePixel.g;
                     blue += sourcePixel.b;
                     alpha += sourcePixel.a;
                  }
               }
               if (totalPixels > 0) {
                  TexturePixel color(red / totalPixels, green / totalPixels, blue / totalPixels,
                                     alpha / totalPixels);
                  int stopY = qMin(y + height, size.height());
                  for (int ypos = qMax(y, 0); ypos < stopY; ypos++) {
                     int linestart = ypos * size.width();
                     int stopX = qMin(x + width, size.width());
                     for (int xpos = qMax(x, 0); xpos < stopX; xpos++) {
                        destimage[linestart + xpos] = color;
                     }
                  }
               }
            }
            x += width;
         }
      }
      y += height;
   }
}

File: generators/pixelate.js
const generator = {
  apiVersion: 1,
  name: "Pixelate",
  description: "Reduces an image to a grid of rectangular colour blocks.",
  type: "filter",
  inputs: ["Image"],

  // Each entry creates a control in the node settings panel.
  settings: [
    {
      id: "width",
      type: "real",
      name: "Block width (%)",
      description: "Width of each block; 0% leaves the input unchanged.",
      default: 5,
      min: 0,
      max: 50,
      group: "size",
    },
    {
      id: "height",
      type: "real",
      name: "Block height (%)",
      description: "Height of each block; 0% leaves the input unchanged.",
      default: 5,
      min: 0,
      max: 50,
      group: "size",
    },
    {
      id: "offsetx",
      type: "real",
      name: "Horizontal offset (%)",
      description: "Moves the block grid horizontally across the texture.",
      default: 0,
      min: -50,
      max: 50,
    },
    {
      id: "offsety",
      type: "real",
      name: "Vertical offset (%)",
      description: "Moves the block grid vertically across the texture.",
      default: 0,
      min: -50,
      max: 50,
    },
  ],

  // This function runs whenever the application renders the node.
  generate(size, settings, output, inputs) {
    // Step 1: begin with a transparent output. A filter has nothing to draw when
    // its Image input is disconnected, so the transparent image is returned.
    const outputPixels = output.data;
    outputPixels.fill(0);

    const sourceImage = inputs.Image;
    if (!sourceImage) return;

    const sourcePixels = sourceImage.data;
    const sourceStride = sourceImage.stride;
    const outputStride = output.stride;
    const textureWidth = size.width;
    const textureHeight = size.height;

    // Step 2: a zero size disables the effect. This makes the zero offered by the
    // settings panel useful and avoids silently turning it into a one-pixel block.
    if (settings.width <= 0 || settings.height <= 0) {
      outputPixels.set(sourcePixels);
      return;
    }

    // Convert positive percentages into pixel measurements. Very small positive
    // values still need a one-pixel block so that both loops always advance.
    const blockWidth = Math.max(1, Math.trunc(settings.width * textureWidth / 100));
    const blockHeight = Math.max(1, Math.trunc(settings.height * textureHeight / 100));
    const horizontalOffset = Math.trunc(settings.offsetx * textureWidth / 100);
    const verticalOffset = Math.trunc(settings.offsety * textureHeight / 100);

    // Step 3: find the block boundary at or immediately before the visible origin.
    // Keeping it congruent with the selected offset makes an offset of zero begin
    // exactly at (0, 0), even when a block does not divide the texture evenly.
    const firstBlockLeft = horizontalOffset
      - Math.ceil(horizontalOffset / blockWidth) * blockWidth;
    const firstBlockTop = verticalOffset
      - Math.ceil(verticalOffset / blockHeight) * blockHeight;

    for (
      let blockTop = firstBlockTop;
      blockTop < textureHeight;
      blockTop += blockHeight
    ) {
      // Skip rows of blocks that end before reaching the visible texture.
      if (blockTop + blockHeight <= 0) continue;

      for (
        let blockLeft = firstBlockLeft;
        blockLeft < textureWidth;
        blockLeft += blockWidth
      ) {
        // Skip blocks that end before reaching the visible texture.
        if (blockLeft + blockWidth <= 0) continue;

        // Step 4: add all colours in this block. Sampling wraps around the source
        // edges, so an offset block can use pixels from both sides of the image.
        let premultipliedRedTotal = 0;
        let premultipliedGreenTotal = 0;
        let premultipliedBlueTotal = 0;
        let alphaTotal = 0;
        let sampledPixels = 0;

        for (let sampleY = blockTop; sampleY < blockTop + blockHeight; ++sampleY) {
          // JavaScript keeps a negative sign after %, so add the texture size when
          // necessary to obtain a valid source row from 0 to textureHeight - 1.
          let sourceY = sampleY % textureHeight;
          if (sourceY < 0) sourceY += textureHeight;

          for (let sampleX = blockLeft; sampleX < blockLeft + blockWidth; ++sampleX) {
            let sourceX = sampleX % textureWidth;
            if (sourceX < 0) sourceX += textureWidth;

            const sourceOffset = sourceY * sourceStride + sourceX * 4;
            const alpha = sourcePixels[sourceOffset + 3];
            premultipliedRedTotal += sourcePixels[sourceOffset] * alpha;
            premultipliedGreenTotal += sourcePixels[sourceOffset + 1] * alpha;
            premultipliedBlueTotal += sourcePixels[sourceOffset + 2] * alpha;
            alphaTotal += alpha;
            ++sampledPixels;
          }
        }

        // RGB is weighted by alpha, so invisible colours cannot tint a translucent
        // block. Dividing by total alpha converts it back to ordinary straight RGBA.
        const averageRed = alphaTotal > 0
          ? Math.round(premultipliedRedTotal / alphaTotal)
          : 0;
        const averageGreen = alphaTotal > 0
          ? Math.round(premultipliedGreenTotal / alphaTotal)
          : 0;
        const averageBlue = alphaTotal > 0
          ? Math.round(premultipliedBlueTotal / alphaTotal)
          : 0;
        const averageAlpha = Math.floor(alphaTotal / sampledPixels);

        // Step 5: clip the block to the visible texture, then give every visible
        // pixel the average colour calculated above.
        const visibleLeft = Math.max(0, blockLeft);
        const visibleRight = Math.min(textureWidth, blockLeft + blockWidth);
        const visibleTop = Math.max(0, blockTop);
        const visibleBottom = Math.min(textureHeight, blockTop + blockHeight);

        for (let y = visibleTop; y < visibleBottom; ++y) {
          let outputOffset = y * outputStride + visibleLeft * 4;

          for (let x = visibleLeft; x < visibleRight; ++x) {
            outputPixels[outputOffset] = averageRed;
            outputPixels[outputOffset + 1] = averageGreen;
            outputPixels[outputOffset + 2] = averageBlue;
            outputPixels[outputOffset + 3] = averageAlpha;
            outputOffset += 4;
          }
        }
      }
    }
  },
};
